Python/Product/Analysis/LanguageServer/ParseQueue.cs

```diff
@@ -32,181 +32,224 @@ class ParseQueue {
+            }
 
-                if (buffers.Count == 1) {
-                    entry.UpdateTree(buffers.First().Value.Ast, cookie);
-                    needAbort = false;
-                    return cookie;
-                }
+            var cookie = new VersionCookie(buffers);
 
-                var tree = new PythonAst(buffers.Values.Select(v => v.Ast));
-                entry.UpdateTree(tree, cookie);
-                needAbort = false;
-                return cookie;
-            } finally {
-                if (needAbort) {
-                    AbortParsingTree(entry);
-                }
+            if (buffers.Count == 1) {
+                tree = buffers.First().Value.Ast;
+            } else {
+                tree = new PythonAst(buffers.Values.Select(v => v.Ast));
             }
+
+            return new StaticPythonParse(tree, cookie);
         }
 
         public Dictionary<string, DiagnosticSeverity> TaskCommentMap { get; set; }
 
         public DiagnosticSeverity InconsistentIndentation { get; set; }
 
-        private void ParsePython(
+        sealed class ParseTask {
+            private readonly ParseQueue _queue;
+            private readonly IDocument _document;
+            private readonly PythonLanguageVersion _languageVersion;
+
+            private readonly IPythonParse _parse;
+
+            private readonly TaskCompletionSource<IAnalysisCookie> _tcs;
+            private Task<IAnalysisCookie> _previous;
+
+            // State transitions:
+            //  UNSTARTED -> QUEUED     when Start() called and _previous is set
+            //  UNSTARTED -> DISPOSED   when DisposeIfNotStarted() is called
+            //  UNSTARTED -> STARTED    when Start() is called
+            //  QUEUED    -> STARTED    when Start(QUEUED) is called
+            //  STARTED   -> DISPOSED   when task completes
+            // Note that calling Dispose() only has an effect on a task that
+            // has not been started or queued.
+            private const int UNSTARTED = 0;
+            private const int QUEUED = 1;
+            private const int STARTED = 2;
+            private const int DISPOSED = 3;
+            private int _state = UNSTARTED;
+
+            public ParseTask(ParseQueue queue, IDocument document, PythonLanguageVersion languageVersion) {
+                _queue = queue;
+                _document = document;
+                _languageVersion = languageVersion;
+
+                _queue._parsingInProgress.Increment();
+                _parse = (_document as IPythonProjectEntry)?.BeginParse();
+
+                _tcs = new TaskCompletionSource<IAnalysisCookie>();
+            }
+
+            public Task<IAnalysisCookie> Task => _tcs.Task;
+
+            public void DisposeIfNotStarted() {
+                if (Interlocked.CompareExchange(ref _state, DISPOSED, UNSTARTED) == UNSTARTED) {
+                    _parse?.Dispose();
```

Comment:
This code is unreachable. `parsing.AddOrUpdate` will always return current task,  `Start` is always called before `DisposeIfNotStarted` and it changes `_state` to either `QUEUED` or `STARTED` before any exception can occur.

Reply:
There are a few ways that it could be reached (all very exceptional), so I'd rather leave it there. But I have streamlined the transitions a bit.

Reply:
> There are a few ways that it could be reached (all very exceptional), so I'd rather leave it there. 

Sure, but can you give an example? I can see it happening only if there is a bug in `ConcurrentDictionary` that it either crashes with exception or returns a value that is different from what has been added or updated.